### PR TITLE
feat: support for `create external table excel <options>`

### PR DESCRIPTION
**Pull request by hailelagi**
closes https://github.com/GlareDB/glaredb/issues/2644, summary of changes:
- implements the `create external table excel <options>` sql command by adding a "streaming table" interface that maps excel rows/columns into a `RecordBatch`.

this will likely be superseded with a followup after the `TableProvider` is implemented for (https://github.com/GlareDB/glaredb/issues/2645) to allow querying on tables and it's missing support for streaming multiple sheets from a worksheet, it's also not super optimal quite yet, but happy to iterate on it on this pr or a followup.

### CODE REVIEW

File: crates/datasources/src/excel/errors.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+use calamine;
+use datafusion::arrow::error::ArrowError;
+use datafusion::error::DataFusionError;
+use datafusion_ext::errors::ExtensionError;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ExcelError {
+    #[error("Failed to load XLSX: {msg}")]
+    Load { msg: String },
```
Comment (tychoish) [inline crates/datasources/src/excel/errors.rs:9]: I'd have this be `Load(String)`

File: crates/datasources/src/excel/errors.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    #[error("Failed to create record batch: {0}")]
+    CreateRecordBatch(#[from] ArrowError),
+    #[error("Failed to open workbook: {0}")]
+    OpenWorkbook(#[from] calamine::XlsxError),
```
Comment (tychoish) [inline crates/datasources/src/excel/errors.rs:13]: Have generic `CalamineError` and `ArrowError`, with specifics for these "create record batch and Open Workbook cases"

File: crates/datasources/src/excel/stream.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+use datafusion::physical_plan::{RecordBatchStream, SendableRecordBatchStream};
+use futures::{Stream, StreamExt};
+
+use crate::excel::errors::ExcelError;
+
+pub type SendableSheetStream = Pin<Box<dyn Stream<Item = Result<RecordBatch, ExcelError>> + Send>>;
+
+pub struct ExcelStream {
+    schema: Arc<Schema>,
+    stream: Pin<Box<dyn Stream<Item = Result<RecordBatch, ExcelError>> + Send>>,
```
Comment (tychoish) [inline crates/datasources/src/excel/stream.rs:19]: does the `SendableRecordBatchStream` alias cover this type?

File: crates/datasources/src/excel/table.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+pub async fn excel_streaming_table(
+    _store_access: GenericStoreAccess,
+    source_url: DatasourceUrl,
+    sheet_name: Option<&str>,
+    has_header: bool,
+    infer_schema_length: usize,
+) -> Result<Arc<dyn TableProvider>, ExcelError> {
+    let mut workbook: Xlsx<_> = match source_url {
+        DatasourceUrl::File(path) => open_workbook(path)?,
+        DatasourceUrl::Url(_) => todo!("remote files not supported yet"),
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:25]: we've talked about this in other contexts, but I think it's pretty straight forward at this point.

File: crates/datasources/src/excel/table.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    };
+
+    // for now, this only infers table data from only the first sheet in a workbook
+    // when read_excel gets vectorised to real columns this should be updated
+    let sheet = sheet_name.map(Cow::Borrowed).unwrap_or_else(|| {
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:30]: I'm not sure that we're getting much out of the `Cow` wrapping, but I suppose it's not hurting anyone here.

File: crates/datasources/src/excel/table.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        let first = sheets.first().expect("sheet is not empty");
+        Cow::Owned(first.clone())
+    });
+
+    let working_sheet_range = match workbook.worksheet_range(&sheet) {
+        Ok(r) => Ok(r),
+        Err(_) => Err(ExcelError::Load {
+            msg: "Failed to open .xlsx file.".to_owned(),
+        }),
+    };
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:41]: can't we replace the match here, with `.map_err(|e| <...>)?`

File: crates/datasources/src/excel/table.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    let range = working_sheet_range?;
+    let schema = excel::infer_schema(&range, has_header, infer_schema_length)?;
+
+    // todo: expand this handle multiple sheets
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:46]: capitalize TODO; 

I think there's a lot of possible options here for dealing with multiple sheets, but the one that sort of makes the most sense to me is to treat sheets as tables and workbooks as databases. So we might have a different implementation that would handle that.

File: crates/datasources/src/excel/table.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    let mut streams = Vec::<Arc<(dyn PartitionStream + 'static)>>::with_capacity(1);
+
+    streams.push(Arc::new(ExcelSheetPartitionStream::new(
+        Arc::new(schema.clone()),
+        Box::pin(futures::stream::iter(vec![
+            xlsx_sheet_value_to_record_batch(range, has_header, infer_schema_length),
+        ])),
+    )));
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:54]: why not make this just `vec![Arc::new(ExcelSheetPartitionStream::new()]`?

File: crates/datasources/src/excel/table.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    let table_stream = StreamingTable::try_new(Arc::new(schema.clone()), streams);
+
+    match table_stream {
+        Ok(streaming_table) => Ok(Arc::new(streaming_table)),
+        Err(_) => Err(ExcelError::Load {
+            msg: "Failed to map to an excel table".to_owned(),
+        }),
+    }
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:63]: you can wrap the table_stream in an `Arc` on line 56, and do this with a `.map_err()?`

File: testdata/sqllogictests/xlsx.slt [hunk 10]
```diff
@@ -57,3 +57,21 @@ select "HEADING" from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xl
...
 statement error
 select * from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xlsx', sheet_name => 'other', infer_rows => -1);
+
+# create external table
+
+statement ok
+create schema 'xlsx';
+
+statement ok
+set search_path = 'xlsx';
```
Comment (tychoish) [inline testdata/sqllogictests/xlsx.slt:67]: is there a reason to do this in a different schema?

File: testdata/sqllogictests/xlsx.slt [hunk 11]
```diff
@@ -57,3 +57,21 @@ select "HEADING" from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xl
...
+
+statement ok
+create external table basic_report from excel options( location='./testdata/xlsx/userdata1.xlsx', file_type = 'xlsx');
+
+# implicitly create
```
Comment (tychoish) [inline testdata/sqllogictests/xlsx.slt:72]: this seems explicit to me?

File: crates/datasources/src/excel/table.rs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+use datafusion::physical_plan::streaming::PartitionStream;
+
+use super::xlsx_sheet_value_to_record_batch;
+use crate::common::url::DatasourceUrl;
+use crate::excel;
+use crate::excel::errors::ExcelError;
+use crate::excel::stream::ExcelSheetPartitionStream;
+use crate::object_store::generic::GenericStoreAccess;
+
+pub async fn excel_streaming_table(
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:16]: so I really like the streaming table model that DF provides, _but_ they are pretty simple, and offer the greatest amount of benefit when there are multiple files (streams) or very large files that can't be in memory for whatever reason. 

In the case of these excel files, there's only ever going to be one file, and I think all or most of it has to be in memory to execute the query, so we might be able to get better results by:
writing our own query plan manually, take advantage of pushdown operations, and support more optimization in the future.

File: testdata/sqllogictests/xlsx.slt [hunk 13]
```diff
@@ -57,3 +57,21 @@ select "HEADING" from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xl
...
+create schema 'xlsx';
+
+statement ok
+set search_path = 'xlsx';
+
+statement ok
+create external table basic_report from excel options( location='./testdata/xlsx/userdata1.xlsx', file_type = 'xlsx');
+
+# implicitly create 
+statement ok
```
Comment (tychoish) [inline testdata/sqllogictests/xlsx.slt:73]: seems like it'd be ok to test some of the error cases/ or edge cases just to demonstrate that we get errors (safely) when provided with bad inputs.

**Review by tychoish — COMMENTED** [review]
comments.

File: crates/datasources/src/excel/table.rs (hunk 12 above)
Comment (hailelagi) [reply crates/datasources/src/excel/table.rs:16]: I kept going back and forth on the mapping here and couldn't decide, even if there were multiple sheets supported it'd all be read into memory anyway by calamite, so yes agree that extending at the query plan makes sense.

File: crates/datasources/src/excel/stream.rs (hunk 3 above)
Comment (hailelagi) [reply crates/datasources/src/excel/stream.rs:19]: I believe it _should_ given:
```rust
impl From<ExcelError> for DataFusionError {
    fn from(e: ExcelError) -> Self {
        DataFusionError::Execution(e.to_string())
    }
}
```

File: crates/datasources/src/excel/table.rs (hunk 7 above)
Comment (hailelagi) [reply crates/datasources/src/excel/table.rs:46]: yupI was thinking this too! will likely follow this up with this bit in separate pr

File: testdata/sqllogictests/xlsx.slt (hunk 10 above)
Comment (hailelagi) [reply testdata/sqllogictests/xlsx.slt]: not really

File: crates/datasources/src/excel/table.rs [hunk 14]
```diff
@@ -4,14 +4,17 @@ use std::sync::Arc;
 use calamine::{open_workbook, Reader, Xlsx};
 use datafusion::datasource::streaming::StreamingTable;
 use datafusion::datasource::TableProvider;
+use datafusion_ext::errors::ExtensionError;
 use datafusion::physical_plan::streaming::PartitionStream;
+// use datafusion::datasource::object_store::ObjectStoreUrl;
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:9]: I would delete this

File: crates/datasources/src/excel/table.rs [hunk 15]
```diff
@@ -22,7 +25,16 @@ pub async fn excel_streaming_table(
 ) -> Result<Arc<dyn TableProvider>, ExcelError> {
     let mut workbook: Xlsx<_> = match source_url {
         DatasourceUrl::File(path) => open_workbook(path)?,
-        DatasourceUrl::Url(_) => todo!("remote files not supported yet"),
+        DatasourceUrl::Url(_) => {
+            /*
+            let store = store_access.create_store().unwrap();
+            let buffer = object_store::buffered::BufReader::with_capacity(store, &meta, infer_schema_length);
+            let sheets = calamine::open_workbook_auto_from_rs(buffer).unwrap();
+             */
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:33]: is this temporary or should we delete it?

File: crates/datasources/src/excel/table.rs [hunk 16]
```diff
@@ -22,7 +25,16 @@ pub async fn excel_streaming_table(
 ) -> Result<Arc<dyn TableProvider>, ExcelError> {
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:25]: It wouldn't be absurd to just convert this type into `ExtensionError` or `DataFusionError` at this point rather than at the call site.

File: crates/datasources/src/excel/table.rs [hunk 17]
```diff
@@ -33,32 +45,26 @@ pub async fn excel_streaming_table(
...
-        Ok(r) => Ok(r),
-        Err(_) => Err(ExcelError::Load {
-            msg: "Failed to open .xlsx file.".to_owned(),
-        }),
-    };
-
-    let range = working_sheet_range?;
+    let range = workbook
+        .worksheet_range(&sheet)
+        .map_err(|e| Into::<ExcelError>::into(e))?;
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:50]: You might be able to get `Ok(workbook.worksheet_range(&sheet)?)` to get the conversion, right? I mgiht be wrong about this case.

File: crates/datasources/src/excel/table.rs [hunk 18]
```diff
@@ -33,32 +45,26 @@ pub async fn excel_streaming_table(
...
-    let mut streams = Vec::<Arc<(dyn PartitionStream + 'static)>>::with_capacity(1);
-
-    streams.push(Arc::new(ExcelSheetPartitionStream::new(
-        Arc::new(schema.clone()),
-        Box::pin(futures::stream::iter(vec![
-            xlsx_sheet_value_to_record_batch(range, has_header, infer_schema_length),
-        ])),
-    )));
+    // TODO: expand this handle multiple sheets
+    let stream: Vec<Arc<(dyn PartitionStream + 'static)>> =
```
Comment (tychoish) [inline crates/datasources/src/excel/table.rs:54]: I don't think you need to specify the type in the vector declaration `Vec<_>` should work.

File: crates/datasources/src/excel/table.rs (hunk 12 above)
Comment (tychoish) [reply crates/datasources/src/excel/table.rs:19]: well or just providing an implementation of `TableProvider` and `ExecutionPlan` rather than using the streaming version. I think deep down the code doesn't change just gets moved around a bit,

File: crates/datasources/src/excel/table.rs (hunk 15 above)
Comment (hailelagi) [reply crates/datasources/src/excel/table.rs:33]: temporary, was an experiment that I intended to come back to, deleting.

**Comment by hailelagi** [comment]
phew! ch-cha-changes! @tychoish like we discussed the "streaming" table idea didn't quite map as nicely so it's currently a partial impl of `TableProvider` and a straightforward `ExecutionPlan`, notably it's still missing lots of features and improvements like `INSERT INTO`, `COPY` & `FILTER` or `WHERE` predicates but those can be added on in a separate issue hopefully. This also includes an `ObjectStore` interface to abstract away local vs remote file handling.

File: crates/datasources/src/excel/stream.rs (hunk 3 above)
Comment (hailelagi) [reply crates/datasources/src/excel/stream.rs:17]: it no longer does.

File: crates/datasources/src/excel/mod.rs [hunk 19]
```diff
@@ -1,89 +1,109 @@
...
+                })
+            }
+
+            DatasourceUrl::Url(_) => {
+                let store = store_access.create_store()?;
+                let list = store_access
+                    .list_globbed(&store, source_url.path().as_ref())
+                    .await?;
+
+                if list.is_empty() {
```
Comment (tychoish) [inline crates/datasources/src/excel/mod.rs:59]: should we error if there's more than one file? do we have a good way of handling multiple files?

File: crates/datasources/src/excel/mod.rs [hunk 20]
```diff
@@ -143,27 +163,69 @@ fn xlsx_sheet_value_to_record_batch(
...
+        .iter()
+        .map(|col_name| {
+            let set = col_types.entry(col_name).or_insert_with(|| {
+                let mut set = HashSet::new();
+                set.insert(DataType::Utf8);
+                set
+            });
+
+            let mut dt_iter = set.iter().cloned();
+            let dt = dt_iter.next().unwrap_or(DataType::Utf8);
```
Comment (tychoish) [inline crates/datasources/src/excel/mod.rs:210]: can't we `into_iter` and avoid the mutable/extra declaration? 

is there a case where the datatype isn't string/utf8 here?

File: testdata/sqllogictests/xlsx.slt [hunk 21]
```diff
@@ -57,3 +57,37 @@ select "HEADING" from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xl
 # negatives are not allowed for infer_rows
 statement error
 select * from read_excel('file://${PWD}/testdata/xlsx/multiple_sheets.xlsx', sheet_name => 'other', infer_rows => -1);
+
+
+# create external table
```
Comment (tychoish) [inline testdata/sqllogictests/xlsx.slt:62]: can we try tests/external tables with different sheet names and make sure that we get different sheets when specified?)

**Review by tychoish — COMMENTED** [review]
this is looking really good!

File: crates/datasources/src/excel/mod.rs (hunk 19 above)
Comment (hailelagi) [reply crates/datasources/src/excel/mod.rs:59]: yup an err makes sense for now if there's more than one file, likely this bit would get refactored when adding the machinery for `create database excel`

File: crates/datasources/src/excel/mod.rs (hunk 20 above)
Comment (hailelagi) [reply crates/datasources/src/excel/mod.rs]: > is there a case where the datatype isn't string/utf8 here?
added better err handling for this at the call site in `infer_value_type`, the `unwrap_or` here seems mostly to satisfy iter semantics returning an Option

File: crates/datasources/src/excel/mod.rs (hunk 19 above)
Comment (tychoish) [reply crates/datasources/src/excel/mod.rs:59]: my initial expectation is that "create database excel" would still be only one file: each and sheet would be it's own table...

**Review by tychoish — APPROVED** [review]

File: crates/datasources/src/excel/mod.rs (hunk 19 above)
Comment (hailelagi) [reply crates/datasources/src/excel/mod.rs:59]: ah... alright then.

**Review by universalmind303 — APPROVED** [review]
